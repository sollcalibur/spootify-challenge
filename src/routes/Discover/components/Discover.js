import React, { Component } from 'react';
import _ from 'lodash';

import { connect } from 'react-redux';
import { requestNewReleases, requestFeaturedPlaylists, requestCategories } from "../../../store/DiscoverActions";
import CoreLayout from '../../../common/layouts/CoreLayout';

import DiscoverBlock from './DiscoverBlock/components/DiscoverBlock';
import '../styles/_discover.scss';

class Discover extends Component {
  componentDidMount() {
    this.props.requestNewReleases();
    this.props.requestFeaturedPlaylists();
    this.props.requestCategories();
  }

  render() {
    const { newReleases, featuredPlaylists, categories } = this.props;
    const isLoaded = !_.isEmpty(newReleases) && !_.isEmpty(featuredPlaylists) && !_.isEmpty(categories)
    const content = isLoaded ? (
      <div className="discover">
        <DiscoverBlock text="RELEASED THIS WEEK" id="released" data={newReleases.albums.items} />
        <DiscoverBlock text="FEATURED PLAYLISTS" id="featured" data={featuredPlaylists.playlists.items} />
        <DiscoverBlock text="BROWSE" id="browse" data={categories.categories.items} imagesKey="icons" />
      </div>
    ) : null;

    return (
      <CoreLayout>
        {content}
      </CoreLayout>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return {
    requestNewReleases: () => dispatch(requestNewReleases()),
    requestFeaturedPlaylists: () => dispatch(requestFeaturedPlaylists()),
    requestCategories: () => dispatch(requestCategories()),
  };
}

function mapStateToProps(state) {
  return {
    newReleases: state.discover.newReleases,
    featuredPlaylists: state.discover.featuredPlaylists,
    categories: state.discover.categories
  };
}

export default connect(
  mapStateToProps, mapDispatchToProps
)(Discover);