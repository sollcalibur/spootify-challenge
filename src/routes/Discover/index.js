import React from 'react';

import _ from 'lodash';
import { getParamValues } from '../../utils';

import CoreLayout from '../../common/layouts/CoreLayout';
import Discover from './components/Discover';

export default class DiscoverContainer extends React.Component {
    componentDidMount() {
        const { history, location } = this.props;
        try {
            if (_.isEmpty(location.hash)) {
                return history.push('/discover');
            }
            
            const access_token = getParamValues(location.hash);
            const expiryTime = 0;

            localStorage.removeItem("params");
            localStorage.removeItem("expiry_time");

            localStorage.setItem("params", JSON.stringify(access_token));
            localStorage.setItem("expiry_time", expiryTime);

            history.push('/discover');
        } catch (error) {
            history.push('/');
        }
    }

    render() {
        return (
            <CoreLayout>
                <Discover></Discover>
            </CoreLayout>
        );
    }
}
