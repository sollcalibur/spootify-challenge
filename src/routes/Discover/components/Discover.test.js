import React from 'react';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Discover from './Discover';

const mockStore = configureStore([thunk]);

describe('Discover component', () => {
    let store;
    let component;


    beforeEach(() => {
        store = mockStore({
            discover: {
                newReleases: {},
                featuredPlaylists: {},
                categories: {},
            }
        });

        store.dispatch = jest.fn();

        component = renderer.create(
            <Provider store={store}>
                <Discover />
            </Provider>
        );
    });

    it('should render with given initial state from Redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should dispatch 3 actions upon mounting the component', () => {
        expect(store.dispatch).toHaveBeenCalledTimes(3);
    });

    it('should render contents when all states are not empty', () => {
        let nonEmptyStore;
        let nonEmptyComponent;

        nonEmptyStore = mockStore({
            discover: {
                newReleases: {
                    albums: {
                        items: [
                            {
                                "name": "Pharrell Williams",
                                images: [
                                    {
                                        "height": 640,
                                        "url": "https://i.scdn.co/image/e6b635ebe3ef4ba22492f5698a7b5d417f78b88a",
                                        "width": 640
                                    }
                                ],
                            },
                        ]
                    }
                },
                featuredPlaylists: {
                    playlists: {
                        items: [
                            {
                                "name": "Pharrell Williams",
                                images: [
                                    {
                                        "height": 640,
                                        "url": "https://i.scdn.co/image/e6b635ebe3ef4ba22492f5698a7b5d417f78b88a",
                                        "width": 640
                                    }
                                ],
                            },
                        ]
                    }
                },
                categories: {
                    categories: {
                        items: [
                            {
                                "name": "Pharrell Williams",
                                icons: [
                                    {
                                        "height": 640,
                                        "url": "https://i.scdn.co/image/e6b635ebe3ef4ba22492f5698a7b5d417f78b88a",
                                        "width": 640
                                    }
                                ],
                            },
                        ]
                    }
                },
            }
        });

        nonEmptyComponent = renderer.create(
            <Provider store={nonEmptyStore}>
                <Discover />
            </Provider>
        );

        const testInstance = nonEmptyComponent.root;
        const element = testInstance.findByType("div");
        expect(element.props.className.includes("discover")).toBe(true);
    });
});