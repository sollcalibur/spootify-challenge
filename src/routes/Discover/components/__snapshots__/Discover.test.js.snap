// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discover component should render with given initial state from Redux store 1`] = `null`;
